### DISCUSSION TITLE
Create Hatch

### DISCUSSION THREAD

**Discussion by thaiguer** [DomCR/ACadSharp#466] [Q&A] [answered]
Hello,
Using the following code, I managed to create a Line and a Circle. But the Hatch isn't visible in the output file.

```
CadDocument newDoc = new CadDocument();

var line = new Line
{
    StartPoint = new CSMath.XYZ(0,0,0),
    EndPoint = new CSMath.XYZ(0,5,0)
};

var circle = new Circle
{
    Center = new CSMath.XYZ(5, 5, 0),
    Radius = 5
};

Hatch.BoundaryPath boundaryPath = new Hatch.BoundaryPath();
boundaryPath.Entities.Add(circle);

List<Hatch.BoundaryPath> boundaryPaths = new List<Hatch.BoundaryPath>();
boundaryPaths.Add(boundaryPath);

var hatch = new Hatch
{
    IsSolid = true,
    Paths = boundaryPaths,
    Color = Color.ByLayer,
    IsAssociative = false,
    Pattern = HatchPattern.Solid,
    IsInvisible = false
};

newDoc.ModelSpace.Entities.Add(line);
newDoc.ModelSpace.Entities.Add(circle);
newDoc.ModelSpace.Entities.Add(hatch);

using (DwgWriter writer = new DwgWriter(targetFile, newDoc))
{
    writer.Write();
}
```

Take a look at this image from the output file:
![image](https://github.com/user-attachments/assets/02d191e7-3cb5-4132-a08b-7cef98b3d953)

What am I missing in the code to be able to create the Hatch correctly?

Important detail: The Hatch is inside the file, but it is not visible.
![image](https://github.com/user-attachments/assets/6b630e54-fd7d-4772-bc89-de6646f88373)
![image](https://github.com/user-attachments/assets/e634387f-97ea-4169-a88a-78370cd45e82)
![image](https://github.com/user-attachments/assets/ede066b8-8644-4856-8c57-bb330bd23c0a)

**Comment by DomCR** [comment]
Hi @thaiguer 

I'm not sure how Autocad handles the entities in the `BoundaryPath` I think they are only used as a reference, meaning that if you change the entity size the hatch is updated as well, but as far as I know, to draw a Hatch you need to create the edges, adding the entity doesn't do the trick.

So this line `boundaryPath.Entities.Add(circle);` is not creating anything, only referencing (I think).

Hope this helps.

**Reply by DomCR** [reply]
I guess that a useful feature would be to be able to create hatches using closed entities like, circles or polylines, I'll take a look at this.

**Reply by rnilay81** [reply]
@thaiguer  Did you manage to get the Hatch in the output. I am also trying to get the Hatch generated. There are no errors but Hatch is still not visible. If you managed to get the Hatch, it would be great if you  can share some pointers/direction you took to get this resolved.

**Reply by thaiguer** [reply]
> @thaiguer Did you manage to get the Hatch in the output. I am also trying to get the Hatch generated. There are no errors but Hatch is still not visible. If you managed to get the Hatch, it would be great if you can share some pointers/direction you took to get this resolved.

Hello, @rnilay81.
Using only ACadSharp I haven't properly managed to create a usefull hatch.
There is a workaround, that is create the hatch in a normal dwg/dxf file with some software like Autocad/BricsCad, then copy or clone this with ACadSharp to your file.

Workaround:

```
using ACadSharp;
using ACadSharp.IO;

string autoCadFileWithHatch = @"C:\dev\test.dwg";
string targetFile= @"C:\dev\newFile.dwg";

using (DwgReader reader = new DwgReader(autoCadFileWithHatch))
{
    CadDocument doc = reader.Read();
    CadDocument newDoc = new CadDocument();

    foreach (var entity in doc.Entities)
    {
        doc.Entities.Remove(entity);
        newDoc.Entities.Add(entity);

        //Clone method can also be used, instead of Remove then Add
    }

    using (DwgWriter writer = new DwgWriter(targetFile, newDoc))
    {
        writer.Write();
    }
}
```

**Reply by thaiguer** [reply]
> I guess that a useful feature would be to be able to create hatches using closed entities like, circles or polylines, I'll take a look at this.

Hello, @DomCR.
Have you made some progress on this matter?
It would be very helpfull to create hatchs using only ACadSharp.

Or is there a way that I'm not aware of?

**Reply by rnilay81** [reply]
@thaiguer  Thanks a lot for the reply and the workaround. Will give it a try.   

@DomCR  Yes, this hatching feature would be really useful. or  a working example with ACadSharp creating hatches and be  visible on the CAD  software would be really great if we are missing something.

**Reply by DomCR** [reply]
Hi all,

For a quick example of how to create a Hatch you can check the single file examples:

https://github.com/DomCR/ACadSharp/blob/master/src/ACadSharp.Tests/IO/WriterSingleObjectTests.cs

Here you have a few methods:

- `CreateHatch()`
- `CreateHatchPolyline()`
- `CreateCircleHatch()`

By now the only way to create a Hatch is to manually generate the polylines.

**Reply by DomCR** [reply]
About the feature to create Hatch from entities is on hold, sorry about that.

**Comment by rnilay81** [comment]
Thanks for the reply and the example of hatch. Appreciate it.


Regards
Nilay

On Mon, Apr 7, 2025 at 11:39 AM Albert Domenech ***@***.***>
wrote:

> Hi all,
>
> For a quick example of how to create a Hatch you can check the single file
> examples:
>
>
> https://github.com/DomCR/ACadSharp/blob/master/src/ACadSharp.Tests/IO/WriterSingleObjectTests.cs
>
> Here you have a few methods:
>
>    - CreateHatch()
>    - CreateHatchPolyline()
>    - CreateCircleHatch()
>
> By now the only way to create a Hatch is to manually generate the
> polylines.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DomCR/ACadSharp/discussions/466#discussioncomment-12746349>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC75Y3Q3CE3UXAEUF2LRPDT2YIJDJAVCNFSM6AAAAABPGZLP7WVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENZUGYZTIOI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Regards,
Nilay

**Reply by rnilay81** [reply]
I have tried to create the hatch and it is working fine. I even managed to get the non-solid hatches to work. Really great. Thanks Team.

**Comment by thaiguer** [comment, marked as answer]
This discussion was turned into an issue: #467
